### ISSUE TITLE
Importing Jekyll post

### ISSUE THREAD

**Issue by Geobert** [cobalt-org/cobalt.rs#276]
Sub-command to import jekyll posts

Step 1: #289 

Remaining work
- [ ] Recursively walk directory
- [ ] Convert config
- [ ] Copy assets
- [ ] permalink conversion

**Comment by epage** [comment]
RE frontmatter

We're in transition between frontmatter formats.  The public-facing format is closer to Jekyll and is very sloppy (a liquid Object).  The new internal-facing format is more strictly defined.  We need to finish settling on the format (#257) and write an upgrade tool (similar to what you're import will be) before we can make the swap.

What we currently do is load the original format and effectively import it into the new format, see [the code](https://github.com/cobalt-org/cobalt.rs/blob/master/src/legacy/wildwest.rs).

My suggested course of action
1. Create subcommand
2. Write code to [read document](https://github.com/cobalt-org/cobalt.rs/blob/master/src/document.rs#L237) with Jekyll frontmatter format
3. Convert jekyll frontmatter to the modern [frontmatter](https://github.com/cobalt-org/cobalt.rs/blob/master/src/frontmatter.rs)
4. Convert modern frontmatter into legacy frontmatter (effectively the opposite of what we do on load)
5. [Dump the document](https://github.com/cobalt-org/cobalt.rs/blob/master/src/document.rs#L460).

This will require more round trips but it means that when we move away from the legacy format, we're just deleting code rather than rewriting it.

**Comment by epage** [comment]
RE subcommand

I've been meaning to refactor the subcommands into external programs (like cargo).  Feel free to do that as a separate PR if you want.  Eventually I'll get to it though.

**Comment by epage** [comment]
RE Behavior

What do you see being the arguments this needs?  Will this work in place (modify existing files) or will it write it out to a separate directory?

A great source of inspiration is to see how other static site generators handle import.  Long term, I'd like to see multiple import types, including Evernote's xml.

**Comment by Geobert** [comment]
Thanks for your inputs. First, what "RE" means? ^^' I'm French so sometime I'm missing some vocabulary ^^

I'll check on cargo's source to see how to do external program for subcommand then :)

I was thinking of something like:
`cobalt import jekyll -s <directory with jekyll's posts> -d <output dir>`

**Comment by epage** [comment]
RE effectively means "regarding".

That seems reasonable.

**Comment by epage** [comment]
Oh, the problem with `import` as a subcommand is that is used right now for the git integration.  We might need to rename one of them.

**Comment by Geobert** [comment]
or use another one for this purpose, like `convert` ?

**Comment by Geobert** [comment]
@epage : when you say external program for subcommands, you mean another binary? Do you have one in mind? I believe some will stay built-in? So I can try to refactor or do you prefer I write a new program and let you integrate it in cobalt?

**Comment by epage** [comment]
The crate is currently setup with a lib + main.  All of the frontmatter stuff is in the lib, though not all might be public.  A crate can have more than one executable.

There are a lot of complexities to deferring to external commands.  I won't hold you to doing so.

**Comment by Geobert** [comment]
Ok so my code can sit into cobalt repo? Would make my life easier ^^ I've just setup a `struct` for Jekyll's frontmatter atm.

**Comment by epage** [comment]
Yeah, either way I was planning on it being in the cobalt repo.

**Comment by epage** [comment]
Closing in favor of https://github.com/cobalt-org/cobalt-migrate-jekyll/issues
